common/FastJmp.h

```diff
@@ -0,0 +1,57 @@
+/*  PCSX2 - PS2 Emulator for PCs
+ *  Copyright (C) 2002-2021  PCSX2 Dev Team
+ *
+ *  PCSX2 is free software: you can redistribute it and/or modify it under the terms
+ *  of the GNU Lesser General Public License as published by the Free Software Found-
+ *  ation, either version 3 of the License, or (at your option) any later version.
+ *
+ *  PCSX2 is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ *  without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
+ *  PURPOSE.  See the GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along with PCSX2.
+ *  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+#include <cstdint>
+#include <cstddef>
+
+struct fastjmp_buf
+{
+#ifdef _WIN32
+#if defined(_M_ARM64)
+	static constexpr std::size_t BUF_SIZE = 168;
+#elif defined(_M_AMD64)
+	static constexpr std::size_t BUF_SIZE = 240;
+#else
+	static constexpr std::size_t BUF_SIZE = 24;
+#endif
+#else
+#if defined(_M_ARM64)
+	static constexpr std::size_t BUF_SIZE = 168;
+#elif defined(__x86_64__)
+	static constexpr std::size_t BUF_SIZE = 64;
+#else
+	static constexpr std::size_t BUF_SIZE = 24;
+#endif
+#endif
+
+	alignas(16) std::uint8_t buf[BUF_SIZE];
+};
+
+extern "C" {
+#if defined(_WIN32) && !defined(_M_AMD64)
+int __fastcall fastjmp_set(fastjmp_buf* buf);
+__declspec(noreturn) int __fastcall fastjmp_jmp(const fastjmp_buf* buf, int ret);
+#elif !defined(_WIN32) && !defined(__x86_64)
+__attribute__((fastcall)) int fastjmp_set(fastjmp_buf* buf);
+__attribute__((fastcall, noreturn)) void fastjmp_jmp(const fastjmp_buf* buf, int ret);
+#elif defined(_WIN32)
+int fastjmp_set(fastjmp_buf* buf);
+__declspec(noreturn) int fastjmp_jmp(const fastjmp_buf* buf, int ret);
+#else
+int fastjmp_set(fastjmp_buf* buf);
+__attribute__((noreturn)) int fastjmp_jmp(const fastjmp_buf* buf, int ret);
+#endif
```

Comment:
We already define `__fastcall` on non-windows platforms  (and I have a PR open that will avoid the warning spam by defining it to nothing on x86_64) https://github.com/PCSX2/pcsx2/blob/c1fea5bc162f16c840dea23bdb8e1fdd34a01d22/common/Pcsx2Defs.h#L152

If you want to add something there for noreturn (can call it `NORETURN` or `_noreturn` or whatever) then we can avoid 4 different ifdefs for the different platforms

Reply:
Done. I named it `__noreturn` for consistency, but as far as I'm aware double-underscored identifiers are reserved, so we should probably do a mass rename at some point.